router/router.go

```diff
@@ -0,0 +1,153 @@
+	GeographyHandler     http.Handler
+	SearchRoutesEnabled  bool
+	SearchHandler        http.Handler
+	HomepageHandler      http.Handler
+	BabbageHandler       http.Handler
+}
+
+func New(cfg Config) http.Handler {
+
+	router := pat.New()
+
+	middleware := []alice.Constructor{
+		dprequest.HandlerRequestID(16),
+		log.Middleware,
+		securityHandler,
+		healthcheckHandler(cfg.HealthCheckHandler),
+		serverError.Handler,
+		redirects.Handler,
+	}
+
+	alice := alice.New(middleware...).Then(router)
+
+	router.Handle("/", cfg.HomepageHandler)
+
+	router.Handle("/redir/{data:.*}", cfg.AnalyticsHandler)
+	router.Handle("/download/{uri:.*}", cfg.DownloadHandler)
+	router.Handle("/cookies{uri:.*}", cfg.CookieHandler)
+	router.Handle("/datasets/{uri:.*}", cfg.DatasetHandler)
+	router.Handle("/filters/{uri:.*}", cfg.FilterHandler)
+	router.Handle("/filter-outputs/{uri:.*}", cfg.FilterHandler)
+	router.Handle("/feedback{uri:.*}", cfg.FeedbackHandler)
+
+	if cfg.GeographyEnabled {
+		router.Handle("/geography{uri:.*}", cfg.GeographyHandler)
+	}
+
+	if cfg.SearchRoutesEnabled {
+		router.Handle("/search", cfg.SearchHandler)
+	}
+
+	// if the request is for a file go directly to babbage instead of using the allRoutesMiddleware
+	router.MatcherFunc(hasFileExtMatcher).Handler(cfg.BabbageHandler)
+
+	// If it is a known babbage endpoint go directly to babbage instead of using the allRoutesMiddleware
+	router.MatcherFunc(isKnownBabbageEndpointMatcher).Handler(cfg.BabbageHandler)
+
+	// all other requests go through the allRoutesMiddleware to check the page type first
+	allRoutesMiddleware := allRoutes.Handler(map[string]http.Handler{
+		"dataset_landing_page": cfg.DatasetHandler,
+	}, cfg.ZebedeeClient, cfg.ContentTypeByteLimit)
+
+	babbageRouter := router.PathPrefix("/").Subrouter()
+	babbageRouter.Use(allRoutesMiddleware)
+	babbageRouter.PathPrefix("/").Handler(cfg.BabbageHandler)
+
+	return alice
+}
+
+// securityHandler ...
+func securityHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.URL.Path != "/embed" && !strings.HasPrefix(req.URL.Path, "/visualisations/") {
+			w.Header().Set("X-Frame-Options", "SAMEORIGIN")
+		}
+		h.ServeHTTP(w, req)
+	})
+}
+
+// healthcheckHandler uses the provided handler for /health endpoint, and serves any other traffic to the next handler in chain
+func healthcheckHandler(hc func(w http.ResponseWriter, req *http.Request)) func(h http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			if req.URL.Path == "/health" {
+				hc(w, req)
+				return
+			}
+			h.ServeHTTP(w, req)
+		})
+	}
+}
+
+var knownBabbageEndpoints = []string{
+	"/chartconfig",
+	"/chartimage",
+	"/generator",
+	"/visualisations/",
+	"/timeseriestool",
+	"/search",
+	"/file",
+	"/resource",
+	"/ons/",
+	"/file",
+	"/calendar",
+	"/chart",
+	"/embed",
+	"/export",
+	"/hash",
```

Comment:
Minor point but I think vars should be declared at the top of the package before functions and types. It's just a consistency thing.

Reply:
I started with it at the top but it seemed disconnected from where it was used. I decided to put the matchers in their own file. Still within the router package, but a little more separate. Now the var for the matcher is at the top 😄